File: apps/android/core/model/src/test/kotlin/me/meeshy/sdk/model/call/CallRetryPolicyTest.kt
package me.meeshy.sdk.model.call

import com.google.common.truth.Truth.assertThat
import org.junit.Test

/**
 * Behavioural spec for [CallRetryPolicy] — the pure decision of which call end
 * reasons warrant a « Réessayer » (retry) affordance. Parité web
 * `isRetryableCallFailure` / same failure set as the reliability aggregate:
 * only TRANSIENT establishment/drop failures (a retry often recovers those,
 * prod 2026-07-12 ~16% of calls), never normal outcomes (local/remote hangup,
 * missed, rejected).
 */
class CallRetryPolicyTest {

    @Test
    fun `offers retry for transient connection failures`() {
        assertThat(CallRetryPolicy.isRetryable(CallEndReason.Failed("Couldn't establish the call connection"))).isTrue()
        assertThat(CallRetryPolicy.isRetryable(CallEndReason.ConnectionLost)).isTrue()
    }

    @Test
    fun `does not offer retry for normal outcomes`() {
        assertThat(CallRetryPolicy.isRetryable(CallEndReason.Local)).isFalse()
        assertThat(CallRetryPolicy.isRetryable(CallEndReason.Remote)).isFalse()
        assertThat(CallRetryPolicy.isRetryable(CallEndReason.Missed)).isFalse()
        assertThat(CallRetryPolicy.isRetryable(CallEndReason.Rejected)).isFalse()
    }

    @Test
    fun `a null reason is not retryable`() {
        assertThat(CallRetryPolicy.isRetryable(null)).isFalse()
    }
}

File: apps/web/__tests__/lib/call-retry-policy.test.ts
/**
 * Pure decision: which call end reasons warrant offering a « Réessayer »
 * (retry) affordance. Only TRANSIENT establishment/drop failures — a retry
 * genuinely recovers those (prod 2026-07-12: ~16% of calls end in
 * failed/connectionLost, often transient ICE/TURN hiccups). Normal outcomes
 * (completed/missed/rejected) and server-side death (heartbeatTimeout/
 * garbageCollected) are NOT retried.
 */

import { isRetryableCallFailure } from '@/lib/calls/call-retry-policy';

describe('isRetryableCallFailure', () => {
  it('offers retry for transient connection failures', () => {
    expect(isRetryableCallFailure('failed')).toBe(true);
    expect(isRetryableCallFailure('connectionLost')).toBe(true);
  });

  it('does NOT offer retry for normal outcomes', () => {
    expect(isRetryableCallFailure('completed')).toBe(false);
    expect(isRetryableCallFailure('missed')).toBe(false);
    expect(isRetryableCallFailure('rejected')).toBe(false);
  });

  it('does NOT offer retry for server-side death', () => {
    expect(isRetryableCallFailure('heartbeatTimeout')).toBe(false);
    expect(isRetryableCallFailure('garbageCollected')).toBe(false);
  });

  it('does NOT offer retry for a null / unknown reason', () => {
    expect(isRetryableCallFailure(null)).toBe(false);
    expect(isRetryableCallFailure(undefined)).toBe(false);
  });
});
